#3: Secrets support

nicoweidner opened the pull request:
Resolves https://github.com/decodableco/decodable-pipeline-sdk/issues/3

Maybe some rough edges, but it works now. Should not be merged before the corresponding ticket DE-3674 is resolved in the main service repo.

gunnarmorling commented on sdk/src/main/java/co/decodable/sdk/pipeline/DecodableSecret.java:
```diff
@@ -0,0 +1,27 @@
+import co.decodable.sdk.pipeline.internal.DecodableSecretImpl;
+import java.util.Date;
+
+public interface DecodableSecret {
```
Being part of the SDK's public API, this should have JavaDoc on the class-level as well as all public members.

nicoweidner replied:
Added JavaDocs

gunnarmorling commented on sdk/src/main/java/co/decodable/sdk/pipeline/DecodableSecret.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Date;
+
+public interface DecodableSecret {
+  String getValue();
```
Let's just do `value()`, in line with more modern record naming style rather than 1998 JavaBeans conventions.

nicoweidner replied:
I assume this is meant to apply to all property accessors?

gunnarmorling replied:
Yes, exactly.

gunnarmorling commented on sdk/src/main/java/co/decodable/sdk/pipeline/internal/DecodableSecretImpl.java:
```diff
@@ -0,0 +1,94 @@
+  private static class SecretMetadata {
+    private final String name;
+    private final String description;
+    private final Date createTime;
```
Let's use `Instant` an friends rather than legacy temporal types.

nicoweidner replied:
Done

gunnarmorling commented on sdk/src/main/java/co/decodable/sdk/pipeline/internal/DecodableSecretImpl.java:
```diff
@@ -0,0 +1,94 @@
+    }
+    try {
+      this.value = FileUtils.readFileToString(secretFile, StandardCharsets.UTF_8);
+      var objectMapper = new ObjectMapper();
```
Could be static and thus instantiated only once? It's quite costly (but also thread-safe).

nicoweidner replied:
Done

gunnarmorling reviewed (commented):
A few comments inline. Could you share some example code which shows how this is meant to be used? Might even go into a [code snippet](https://inside.java/2022/04/04/sip46/) in the class comment for `DecodableSecret` (see [here](https://github.com/decodableco/decodable-pipeline-sdk/blob/main/sdk/src/main/java/co/decodable/sdk/pipeline/package-info.java#L20-L21) for prior art in the SDK).

gunnarmorling commented:
Ah, one more thing: the `@Incubating` annotation should be put to all new public types.

nicoweidner commented:
> A few comments inline. Could you share some example code which shows how this is meant to be used? Might even go into a [code snippet](https://inside.java/2022/04/04/sip46/) in the class comment for `DecodableSecret` (see [here](https://github.com/decodableco/decodable-pipeline-sdk/blob/main/sdk/src/main/java/co/decodable/sdk/pipeline/package-info.java#L20-L21) for prior art in the SDK).

I added a snippet, though Spotless didn't like the line breaks in code and removed them 🙄

gunnarmorling commented:
> though Spotless didn't like the line breaks in code and removed them

Yeah, you can disable Spotless there (see the referenced class for an
example).

> Message ID: ***@***.***
> .com>
>

nicoweidner commented:
> though Spotless didn't like the line breaks in code and removed them
> Yeah, you can disable Spotless there (see the referenced class for an example).
> […](#)
> Message ID: ***@***.*** .com>

Thanks, disabled spotless. That even fixed CI!

gunnarmorling commented on sdk/src/main/java/co/decodable/sdk/pipeline/DecodableSecret.java:
```diff
@@ -0,0 +1,53 @@
+ * this:
+ *
+ * <p>
+ * {@snippet :
+   var mySecret = DecodableSecret.withName("my-secret");
+   var mySecretValue = mySecret.value(); doStuffWithMySecret(mySecretValue);
+   }
```
Ah, I was hoping for something more realistic, with an actual connector example. Challenge is it would have to be one which we don't support first-class managed in Decodable while still being relatable for many readers. Maybe it's not a good idea, dunno.

nicoweidner replied:
I was afraid you would say that 🙈.

Not sure I really like the idea, though. The secret class doesn't really have anything to do with specific connectors, and we'd also run the risk of having an outdated JavaDoc if something in the connector changes (e.g. required properties). Moreover, we don't have anything about connectors in the SDK, do we? So we'd have to write lower-level Java/Flink code to connect to some imaginary connection, which feels overkill to me (and too much for a JavaDoc here). But I'll check with the connector folks for ideas.
 
If we want to have an actual example, I'd prefer to have a complete one similar to `DataStreamJob`. It would have to be one where it's easy to set up the connector in Decodable though, so people could actually set up the resources and run the example quickly.

gunnarmorling replied:
> It would have to be one where it's easy to set up the connector in Decodable though

Not quite following on that one. One use case for the SDK--and AFAIU, the motivator for secrets support--is to enable usage of connectors which are _not_ supported as managed connectors on Decodable (yet). So I think this is what we somehow should show in docs (be it JavaDoc or elsewhere).

nicoweidner replied:
So then it would only be theoretical code and users would not be able to actually run it anywhere, right (unless they have an external connection of this type and adapt the parameters)?

Do you happen to know an example of a not yet supported connector that would be interesting to use, and some Flink code to go with it? 😇 

Another use case is to combine the connector with some processing or other logic, which would also make sense for supported connectors. But this may be less interesting than not yet supported connectors.

gunnarmorling replied:
> So then it would only be theoretical code and users would not be able to actually run it anywhere, right (unless they have an external connection of this type and adapt the parameters)?

Yes, exactly.

> Do you happen to know an example of a not yet supported connector that would be interesting to use, and some Flink code to go with it? 😇

The [SQL Server CDC connector](https://ververica.github.io/flink-cdc-connectors/master/content/connectors/sqlserver-cdc.html) would be a good example, e.g. like so:

```
SourceFunction<String> sourceFunction = SqlServerSource.<String>builder()
      .hostname("localhost")
      .port(1433)
      .database("inventory")
      .tableList("dbo.items")
      .username("my-sql-server-user")
      .password(DecodableSecret.withName("my-sql-server-password").value())
      .deserializer(new JsonDebeziumDeserializationSchema())
      .build();
```

nicoweidner replied:
Thanks. Put this in the JavaDoc instead of the dummy example.

nicoweidner commented:
@gunnarmorling Have you seen these build failures before? I wonder whether they are transient or have something to do with the `@snippet`

gunnarmorling reviewed (approved):
LGTM, thanks1
